### PR TITLE
Feature/adapt structure for multi organizations

### PR DESCRIPTION
**Pull request by daniels98it**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://vmbruegge60.in.tum.de/.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context

From https://github.com/ls1intum/Artemis/issues/2710 

Artemis is currently been used in projects where an instance is hosted for multiple universities and other educational organizations. Some difficulties occur to provide a meaningful separation between them (ex. Students self registration to courses of their own organization, allow instructors to view only users of the same organization, ..).

To simplify these task, keeping an organized structure, a new **organization** entity could be added and mapped to users and courses in order to organize these as sub-groups.

##### Structure

The proposed solution to this issue is the integration of a new 'Organization' table, where the needed data of multiple organizations can be stored. The organizations are then mapped to users and courses in order to allow the before mentioned separation. 
UMLClassDiagram (1)

##### Usage

Each organization will have an emailPattern column, needed to assign existing , as well as newly registering users, to an organization, without the need of a manually assignment. 

Admins should be able to add organizations to courses, allowing only users/students of the same organizations to register to these.

##### Management

In order to simplify the management of the organizations, a new view will be added for admins to directly add (or remove) users from organizations, as well as creating/updating/removing organizations.

In addition, organizations can be added/removed in the course/user update views directly.

##### Indexing

When artemis is started, the users will be indexed by their email addresses to match currently stored organizations, without the need to manually allocate each user individually to its organization.

Such indexing happens even when the emailPattern of an organization has been updated, updating the users list of the organization.

When a new user is created, it will automatically be assigned to the matching organizations if its email matches the email pattern of the organization.

##### Course registration

The course registration will be carried out in an additive way. In other words, all other course-registration requirements are still required and a matching organization represents an additional requirement on top of the others (if enabled).

### Steps for Testing

1. Log in to Artemis
2. Navigate to Organization management (under Server administration)
3. Test organization managament:
3.1 Create a new organization (users matching the emailPattern should automatically be added)
3.2 Update the organization (users matching the new emailPattern should automatically be added, others not matching removed)
3.3 Remove the organization

4. Click "view" button of an organization to open the detail view
4.1 Try to add a user using the user table
4.2 Try to remove a user using the user table

5. Navigate to course management and edit an active course:
5.1 Add an organization to the course and make sure student registration is enabled
5.2 Save the updated course

6. Navigate to user management and edit a student-user:
6.1 Add the same organization (as previously set in course) to the user
6.2 Save the updated user

7. Using the updated user try to register to the updated course (course-registration.allowed-username-pattern should match too for this step)

#### Server

CourseRepository: 90%
OrganizationRepository: 79%
UserRepository: 78%
OrganizationService: 81%
CourseResource: 91%
OrganizationResource: 97%
UserResource: 87%

#### Client

organization-management-detail.component.ts 70%
organization-management-update.component.ts 89%
organization-management.component.ts 96%
organization-management.route.ts 73%
organization-management.service.ts 89%
organization-selector.component.ts 95%

### Screenshots

Screenshot from 2021-02-24 14-55-21
Screenshot from 2021-02-24 14-55-35
Screenshot from 2021-02-24 14-57-58
Screenshot from 2021-02-24 14-58-20
Screenshot from 2021-02-24 14-58-43

### CODE REVIEW

**Review by airelawaleria — DISMISSED** [review]
tested in the testing session with @marvinchimeka @JohannesWeiss Works as expected

**Review by JohannesWeiss — DISMISSED** [review]
Tested in testing session on ts1, works as intended so far.

**Review by marvinchimeka — DISMISSED** [review]
Tested in testing session 👍

**Review by EneaGore — DISMISSED** [review]
Did not face any issues while testing on ts1.

**Comment by FrankeLukas** [comment]
I took the freedom to add proper breadcrumbs in https://github.com/ls1intum/Artemis/pull/2748/commits/d11a882f4765ddd043194503ed1bca59203a378e.
Before:
grafik
After:
grafik

**Review by FrankeLukas — APPROVED** [review]
Didn't review all code in detail, but the changes work on TS1.
One thing I noticed is that the display of organisations for a single user doesn't look very nice, maybe we can center the text a bit more here:
grafik

**Review by philgabr — APPROVED** [review]
Works as described :+1:

Aforementioned visual issue also occurs for course organizations:
image

File: src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java [hunk 1]
```diff
@@ -318,4 +357,10 @@ default Long countUserInGroup(String groupName) {
     default void updateUserNotificationReadDate(long userId) {
         updateUserNotificationReadDate(userId, ZonedDateTime.now());
     }
+
+    @Query("select user from User user left join fetch user.organizations")
+    List<User> findAllWithEagerOrganizations();
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java:362]: This is not a good query. It might load 15.000 users, we should avoid this

File: src/main/java/de/tum/in/www1/artemis/domain/Course.java [hunk 2]
```diff
@@ -150,6 +151,13 @@
     @JsonIgnoreProperties("course")
     private Set<Exam> exams = new HashSet<>();
 
+    @ManyToMany
+    @JoinTable(name = "course_organization", joinColumns = { @JoinColumn(name = "course_id", referencedColumnName = "id") }, inverseJoinColumns = {
+            @JoinColumn(name = "organization_id", referencedColumnName = "id") })
+    @Cache(usage = CacheConcurrencyStrategy.NONSTRICT_READ_WRITE)
+    @BatchSize(size = 20)
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/domain/Course.java:158]: I would leave this out for now

File: src/main/java/de/tum/in/www1/artemis/domain/Organization.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+    @NotNull
+    @Column(name = "emailPattern")
+    private String emailPattern;
+
+    @ManyToMany
+    @JoinTable(name = "user_organization", joinColumns = { @JoinColumn(name = "organization_id", referencedColumnName = "id") }, inverseJoinColumns = {
+            @JoinColumn(name = "user_id", referencedColumnName = "id") })
+    @Cache(usage = CacheConcurrencyStrategy.NONSTRICT_READ_WRITE)
+    @BatchSize(size = 20)
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/domain/Organization.java:49]: remove BatchSize and add JsonIgnoreProperties

File: src/main/java/de/tum/in/www1/artemis/domain/Organization.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    private Set<User> users = new HashSet<>();
+
+    @ManyToMany
+    @JoinTable(name = "course_organization", joinColumns = { @JoinColumn(name = "organization_id", referencedColumnName = "id") }, inverseJoinColumns = {
+            @JoinColumn(name = "course_id", referencedColumnName = "id") })
+    @Cache(usage = CacheConcurrencyStrategy.NONSTRICT_READ_WRITE)
+    @BatchSize(size = 20)
+    private Set<Course> courses = new HashSet<>();
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/domain/Organization.java:57]: remove BatchSize and add JsonIgnoreProperties

File: src/main/java/de/tum/in/www1/artemis/domain/User.java [hunk 5]
```diff
@@ -116,6 +116,13 @@
     @BatchSize(size = 20)
     private Set<Authority> authorities = new HashSet<>();
 
+    @ManyToMany
+    @JoinTable(name = "user_organization", joinColumns = { @JoinColumn(name = "user_id", referencedColumnName = "id") }, inverseJoinColumns = {
+            @JoinColumn(name = "organization_id", referencedColumnName = "id") })
+    @Cache(usage = CacheConcurrencyStrategy.NONSTRICT_READ_WRITE)
+    @BatchSize(size = 20)
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/domain/User.java:123]: remove BatchSize and add JsonIgnoreProperties

File: src/main/java/de/tum/in/www1/artemis/service/OrganizationService.java [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    /**
+     * Called at application startup, this method tries to assign organizations to
+     * users which do not belong to any. For each user without organization, the email
+     * patterns of the organization are used to find any matching, resulting in the assignment
+     * of the user to the matcher organization.
+     */
+    @PostConstruct
+    public void init() {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/OrganizationService.java:46]: This should only be invoked once.
Also this does not really support multi node environments

Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/OrganizationService.java:46]: I suggest to let the system admin invoke this functionality with a button.

Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/OrganizationService.java:46]: I suggest Artemis shows a warning and the user needs to confirm this operation

File: src/main/java/de/tum/in/www1/artemis/domain/Course.java [hunk 7]
```diff
@@ -150,6 +151,13 @@
     @JsonIgnoreProperties("course")
     private Set<Exam> exams = new HashSet<>();
 
+    @ManyToMany
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/domain/Course.java:154]: please talk to @stefanwaldhauser about the use of `mappedBy`

File: src/main/java/de/tum/in/www1/artemis/repository/CourseRepository.java [hunk 8]
```diff
@@ -72,6 +72,12 @@
     @Query("select distinct course from Course course where (course.startDate is null or course.startDate <= :#{#now}) and (course.endDate is null or course.endDate >= :#{#now}) and course.onlineCourse = false and course.registrationEnabled = true")
     List<Course> findAllCurrentlyActiveNotOnlineAndRegistrationEnabled(@Param("now") ZonedDateTime now);
 
+    @Query("select distinct course from Course course left join fetch course.organizations co where (course.startDate is null or course.startDate <= :#{#now}) and (course.endDate is null or course.endDate >= :#{#now}) and course.onlineCourse = false and course.registrationEnabled = true")
+    List<Course> findAllCurrentlyActiveNotOnlineAndRegistrationEnabledWithOrganizations(@Param("now") ZonedDateTime now);
+
+    @Query("select course from Course course left join fetch course.organizations co where course.id = :#{#courseId}")
+    Course findWithEagerOrganizations(@Param("courseId") long courseId);
```
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/repository/CourseRepository.java:79]: More modern here would be to return Optional<Course> but just a suggestion.

File: src/main/java/de/tum/in/www1/artemis/domain/Organization.java [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+    @Column(name = "logoUrl")
+    private String logoUrl;
+
+    @NotNull
+    @Column(name = "emailPattern")
+    private String emailPattern;
+
+    @ManyToMany
+    @JoinTable(name = "user_organization", joinColumns = { @JoinColumn(name = "organization_id", referencedColumnName = "id") }, inverseJoinColumns = {
```
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/domain/Organization.java:46]: I would be very careful with using @ManyToMany on both sides. Normally it works like this. The entity with @ManyToMany association is responsible for handling this bidirectional association.**So this is the entity that controls if a realtionship exists. Meaning if you remove it from the collection here the relationship will be removed on the database level. Removing the entity from the @MappedBy side has no effect on the databse level**  I have no idea how this works with manyToMany on both sides

In other words https://stackoverflow.com/questions/5460573/hibernate-bidirectional-manytoone-updating-the-not-owning-side-not-working

This is the way it's supposed to work: the owning side (@ManyToMany) is the one which owns the relationship, and which is responsible of managing it in database. It's the responsibility of the developer (i.e. your responsibility) to make sure your object graph is always coherent : if one side of a relationship is modified, then the other side should be modified as well.

If you don't always do this, you must at least make sure to modify the owning side of the relationship, because Hibernate only cares about the owning side..

File: src/main/resources/config/application-prod.yml [hunk 10]
```diff
@@ -132,4 +132,3 @@ info:
             - tutor_assessment_tour: 'Patterns in Software Engineering'
     contact: artemis.in@tum.de                                                                  #default value, can be overridden on the server
     imprint: https://ase.in.tum.de/lehrstuhl_1/component/content/article/179-imprint            #default value, can be overridden on the server
-    test-server: false  # false --> production, true --> test server, --> empty == local
```
Comment (stefanwaldhauser) [inline src/main/resources/config/application-prod.yml:135]: Is that removal intended?

**Review by stefanwaldhauser — CHANGES_REQUESTED** [review]
Did not have time yet to check everything but looks like an important and promising chang! Gj

File: src/main/resources/config/application-prod.yml (hunk 10 above)
Comment (daniels98it) [reply src/main/resources/config/application-prod.yml:135]: Nope.. Will re-add in the next commit

Thanks for noticing :+1:

File: src/main/java/de/tum/in/www1/artemis/repository/CourseRepository.java (hunk 8 above)
Comment (daniels98it) [reply src/main/java/de/tum/in/www1/artemis/repository/CourseRepository.java]: You are right, i missed a 'findWithEagerOrganizations**ElseThrow**' method which returns the Course entity directly.

Added Optional and the additional method in 7507bfa

File: src/main/java/de/tum/in/www1/artemis/domain/Organization.java (hunk 9 above)
Comment (daniels98it) [reply src/main/java/de/tum/in/www1/artemis/domain/Organization.java]: Quite sure I extensively tested it with **manyToMany** on both sides and didn't show inconsistencies. Problem was I wanted to manage the Organizations from both sides (adding users/courses from the Organization side and adding Organizations along with Course/User entities), which is not possible using **mappedBy**...

Anyway I believe that its better to keep sticking to the **mappedBy** solution, so I did some refactoring in e4c6beb to avoid the double-joins.

P.s. The explanation really helped me out, thanks :rocket:

File: src/main/java/de/tum/in/www1/artemis/domain/Course.java (hunk 7 above)
Comment (daniels98it) [reply src/main/java/de/tum/in/www1/artemis/domain/Course.java:153]: Added **mappedBy** with e4c6beb

File: src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java (hunk 1 above)
Comment (daniels98it) [reply src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java]: Removed and replaced by **findAllMatchingEmailPattern**

File: src/main/java/de/tum/in/www1/artemis/domain/Course.java (hunk 2 above)
Comment (daniels98it) [reply src/main/java/de/tum/in/www1/artemis/domain/Course.java]: Done :+1:

File: src/main/java/de/tum/in/www1/artemis/domain/Organization.java (hunk 3 above)
Comment (daniels98it) [reply src/main/java/de/tum/in/www1/artemis/domain/Organization.java]: Done :+1:

File: src/main/java/de/tum/in/www1/artemis/domain/Organization.java (hunk 4 above)
Comment (daniels98it) [reply src/main/java/de/tum/in/www1/artemis/domain/Organization.java:53]: Done :+1:

File: src/main/java/de/tum/in/www1/artemis/domain/User.java (hunk 5 above)
Comment (daniels98it) [reply src/main/java/de/tum/in/www1/artemis/domain/User.java]: Done :+1:

File: src/main/java/de/tum/in/www1/artemis/service/OrganizationService.java (hunk 6 above)
Comment (daniels98it) [reply src/main/java/de/tum/in/www1/artemis/service/OrganizationService.java]: * PostConstruct part (indexing at application startup) has been removed
* Indexing implementation has been adapted to the suggestion (using the new **findAllMatchingEmailPattern** method)
* Indexing now uses 'strict' indexing (removes users which do not match the organization
* I assume the warning for the user is not needed as he/she cannot change his/her own email. Only admins can. (see screenshot)
Screenshot from 2021-03-16 09-06-32

**Review by FrankeLukas — CHANGES_REQUESTED** [review]
Works as advertised. Good work. Only issues we found is we can't delete organisations for some reason.

**Comment by daniels98it** [comment]
Fixed identified issues with d464895

File: src/main/webapp/app/course/manage/course-update.component.ts [hunk 11]
```diff
@@ -368,6 +383,40 @@ export class CourseUpdateComponent implements OnInit {
...
+        /*
+        if (this.course.id !== undefined) {
+            this.organizationService.removeCourseFromOrganization(organization.id!, this.course.id).subscribe(
+                () => {
+                    this.course.organizations = this.course.organizations!.filter((o) => o.id !== organization.id);
+                },
+                (error: HttpErrorResponse) => this.jhiAlertService.error(error.message),
+            );
+        }
+         */
```
Comment (ivanchimeno) [inline src/main/webapp/app/course/manage/course-update.component.ts:418]: I think you can remove the commented code :)

File: src/main/webapp/app/admin/organization-management/organization-management.service.ts [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+     */
+    getOrganizations(): Observable<any> {
+        return this.http.get(this.resourceUrl + '/all');
+    }
+
+    /**
+     * Send GET request to retrieve all organizations with their number of
+     * users and courses
+     */
+    getNumberOfUsersAndCoursesOfOrganizations(): Observable<any> {
```
Comment (ivanchimeno) [inline src/main/webapp/app/admin/organization-management/organization-management.service.ts:25]: Since you know what the service returns, you can change `any` to something more specific like `number` (same for the functions below)

File: src/main/java/de/tum/in/www1/artemis/service/OrganizationService.java [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+    /**
+     * Delete an organization
+     * @param organizationId the id of the organization to delete
+     */
+    public void deleteOrganization(Long organizationId) {
+        final Organization organization = organizationRepository.findOneWithEagerUsersAndCoursesOrElseThrow(organizationId);
+
+        // we make sure to delete all relations before deleting the organization
+        organization.getUsers().forEach(user -> {
+            userRepository.removeOrganizationFromUser(user.getId(), organization);
```
Comment (ivanchimeno) [inline src/main/java/de/tum/in/www1/artemis/service/OrganizationService.java:116]: Not sure here but since it's a database call, would that impact performance if the organization has many users? Is there a way to batch-delete entities?

**Review by ivanchimeno — CHANGES_REQUESTED** [review]
Code looks good overall, just added a few comments

File: src/main/java/de/tum/in/www1/artemis/service/OrganizationService.java (hunk 13 above)
Comment (daniels98it) [reply src/main/java/de/tum/in/www1/artemis/service/OrganizationService.java:116]: Tried to come up with a native query removing the relative relationships but does not seem to work.. 
Unfortunately I did not find an alternative solution and kept these for loops as done by removing courses (`CourseService.java - public void delete(Course course)`)

File: src/main/webapp/app/admin/organization-management/organization-management.service.ts (hunk 12 above)
Comment (daniels98it) [reply src/main/webapp/app/admin/organization-management/organization-management.service.ts:25]: Done with https://github.com/ls1intum/Artemis/pull/2748/commits/ab2264066656190b61efae207cf9a1382a0104d4 but could not come up with a type for `getNumberOfUsersAndCoursesOfOrganizations()` as response from the backend has form `Map<Long, Map<String, Long>>`

File: src/main/webapp/app/course/manage/course-update.component.ts (hunk 11 above)
Comment (daniels98it) [reply src/main/webapp/app/course/manage/course-update.component.ts]: Removed with https://github.com/ls1intum/Artemis/pull/2748/commits/ab2264066656190b61efae207cf9a1382a0104d4

Thanks for noticing :smiley:

**Review by abcdefg30 — APPROVED** [review]
Edit: Wrong account, sorry. :sweat_smile:

**Review by FrankeLukas — APPROVED** [review]
Deleting an organisation works now. ✅ 
One thing I noticed is that you don't have to write the organisation's name when trying to delete it (which is usually the case in other delete dialogues). Do we want to do that here, too?

File: src/main/webapp/app/shared/organization-selector/organization-selector.component.ts [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+        this.organizationService.getOrganizations().subscribe((data) => {
+            this.availableOrganizations = data;
+            if (this.organizations !== undefined) {
+                this.availableOrganizations.forEach((organization) => {
+                    this.organizations.forEach((currentOrganization) => {
+                        if (organization.id === currentOrganization.id) {
+                            this.availableOrganizations = this.availableOrganizations.filter((org) => org.id !== organization.id);
+                        }
+                    });
+                });
```
Comment (iozbeyli) [inline src/main/webapp/app/shared/organization-selector/organization-selector.component.ts:35]: You can use filter and some for this
Like array.filter((item) => fields.some((field) => item[field] === value));

File: src/main/webapp/app/shared/organization-selector/organization-selector.component.ts [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+import { Course } from 'app/entities/course.model';
+
+@Component({
+    selector: 'jhi-organization-selector',
+    templateUrl: './organization-selector.component.html',
+})
+export class OrganizationSelectorComponent implements OnInit {
+    organizations: Organization[];
+    availableOrganizations: Organization[];
+    course: Course;
```
Comment (iozbeyli) [inline src/main/webapp/app/shared/organization-selector/organization-selector.component.ts:16]: Maybe I missed it but do you use this or set it anywhere?

Comment (daniels98it) [reply src/main/webapp/app/shared/organization-selector/organization-selector.component.ts:16]: The component is used in `course-update.component.ts` and `user-management-update.component.ts` as modal and currently assigned organizations are passed using `modalRef.componentInstance.organizations = this.user.organizations;` to the `organizations: Organization[]` property.

Comment (iozbeyli) [reply src/main/webapp/app/shared/organization-selector/organization-selector.component.ts:16]: No, I meant the course variable.

File: src/test/javascript/spec/component/organization/organization-management-detail.component.spec.ts [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+    it('should remove user from organization', fakeAsync(() => {
+        const user1 = new User();
+        user1.id = 11;
+        user1.login = 'userOne';
+        const user2 = new User();
+        user2.id = 12;
+        const user3 = new User();
+        user3.id = 13;
+        organization1.users = [user1, user2, user3];
+        component.organization = organization1;
```
[the remark this answers is not in the document]
Comment (daniels98it) [reply src/test/javascript/spec/component/organization/organization-management-detail.component.spec.ts]: Refactored with 9a38758
Thanks :+1:

File: src/test/javascript/spec/component/organization/organization-management.service.spec.ts [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+        const req = httpMock.expectOne({ method: 'GET' });
+        req.flush(JSON.stringify(elemDefault));
+    });
+
+    it('should return all organizations', async () => {
+        const elem2 = new Organization();
+        elem2.id = 1;
+        elem2.name = 'test2';
+        elem2.shortName = 'test2';
+        const returnElement = [elemDefault, elem2];
```
[the remark this answers is not in the document]
Comment (daniels98it) [reply src/test/javascript/spec/component/organization/organization-management.service.spec.ts]: Refactored with 9a38758
Thanks :+1:

File: src/main/webapp/app/shared/organization-selector/organization-selector.component.ts (hunk 15 above)
Comment (daniels98it) [reply src/main/webapp/app/shared/organization-selector/organization-selector.component.ts:16]: My bad.. You are right (unused) :sweat_smile:

Comment (daniels98it) [reply src/main/webapp/app/shared/organization-selector/organization-selector.component.ts]: Removed with c5025f2
Thanks :+1:

File: src/main/webapp/app/shared/organization-selector/organization-selector.component.ts (hunk 14 above)
Comment (daniels98it) [reply src/main/webapp/app/shared/organization-selector/organization-selector.component.ts]: Thanks, did not know the 'some' function before :slightly_smiling_face: 
Improved with c5025f2 :+1:

**Review by ivohashamov — COMMENTED** [review]
Most functionality seems to work as intended. One thing was noticed during the testing session - whenever users that are registered to a course and are a part of at least one of the course's organizations get removed from the course's organizations they still remain part of the course and can participate in the exercises. I don't know if that's the intended functionality. Perhaps the users could remain part of the course with their submissions intact but be prohibited from participating in further exercises. Or maybe they could be removed altogether from the course.

File: src/main/java/de/tum/in/www1/artemis/service/OrganizationService.java [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+        this.courseRepository = courseRepository;
+    }
+
+    /**
+     * Performs indexing over all users using the email pattern of the provided organization.
+     * Users matching the pattern will be added to the organization.
+     * Users not matching the pattern will be removed from the organization (if contained).
+     * @param organization the organization used to perform the indexing
+     */
+    public void indexing(final Organization organization) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/OrganizationService.java:41]:
```suggestion
    public void index(final Organization organization) {
```

File: src/main/java/de/tum/in/www1/artemis/service/OrganizationService.java [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+     * Update an organization
+     * To avoid removing the currently mapped users and courses of the organization,
+     * these are loaded eagerly and the edited values changed within the loaded entity.
+     * @param organization the organization to update
+     * @return the updated organization
+     */
+    public Organization update(Organization organization) {
+        log.debug("Request to update Organization : {}", organization);
+        boolean indexingRequired = false;
+        Optional<Organization> optionalOldOrganization = organizationRepository.findByIdWithEagerUsersAndCourses(organization.getId());
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/OrganizationService.java:85]:
```suggestion
        var oldOrganization = organizationRepository.findByIdWithEagerUsersAndCoursesElseThrow(organization.getId());
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java [hunk 20]
```diff
@@ -480,7 +484,17 @@ private void validateShortName(Course course) {
     @PreAuthorize("hasAnyRole('USER', 'TA', 'INSTRUCTOR', 'ADMIN')")
     public List<Course> getAllCoursesToRegister() {
         log.debug("REST request to get all currently active Courses that are not online courses");
-        return courseRepository.findAllCurrentlyActiveNotOnlineAndRegistrationEnabled();
+        User user = userRepository.getUserWithGroupsAndAuthoritiesAndOrganizations();
+
+        List<Course> allRegistrable = courseRepository.findAllCurrentlyActiveNotOnlineAndRegistrationEnabledWithOrganizations();
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:489]:
```suggestion
        List<Course> allCoursesToRegister = courseRepository.findAllCurrentlyActiveNotOnlineAndRegistrationEnabledWithOrganizations();
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java [hunk 21]
```diff
@@ -480,7 +484,17 @@ private void validateShortName(Course course) {
...
+        return allRegistrable.stream().filter(course -> {
+            if (course.getOrganizations() != null && course.getOrganizations().size() > 0) {
+                return checkIfUserIsMemberOfCourseOrganizations(user, course);
+            }
+            else {
+                return true;
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:495]: add one comment here so it's easier to understand

**Review by SimonEntholzer — APPROVED** [review]

**Review by FuchsDominik — APPROVED** [review]

**Review by krusche — APPROVED** [review]
we only found some minor code improvements and would like to ask you to fix those in a follow-up PR so that we can finally merge this huge giant ;-)

**Review by FrankeLukas — APPROVED** [review]

**Review by iozbeyli — APPROVED** [review]

**Review by coolchock — APPROVED** [review]

**Review by ivohashamov — APPROVED** [review]

File: src/main/java/de/tum/in/www1/artemis/service/OrganizationService.java (hunk 18 above)
Comment (daniels98it) [reply src/main/java/de/tum/in/www1/artemis/service/OrganizationService.java:41]: Added in https://github.com/ls1intum/Artemis/pull/3102

File: src/main/java/de/tum/in/www1/artemis/service/OrganizationService.java (hunk 19 above)
Comment (daniels98it) [reply src/main/java/de/tum/in/www1/artemis/service/OrganizationService.java:85]: Added in https://github.com/ls1intum/Artemis/pull/3102

File: src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java (hunk 20 above)
Comment (daniels98it) [reply src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:489]: Added in https://github.com/ls1intum/Artemis/pull/3102

File: src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java (hunk 21 above)
Comment (daniels98it) [reply src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:495]: Added in https://github.com/ls1intum/Artemis/pull/3102
